archive/100-theorems-list/88_derangements_formula.lean

```diff
@@ -0,0 +1,451 @@
+def derangement_congr {α β : Type*} (e : α ≃ β) : (derangements α ≃ derangements β) :=
+begin
+  refine subtype_equiv (perm_congr e) _,
+  intro f,
+  simp only [derangements, is_derangement, perm_congr_apply, set.mem_set_of_eq],
+  rw ← not_iff_not,
+  push_neg,
+  split,
+  { rintro ⟨x, hx_fixed⟩,
+    use e x,
+    simp [hx_fixed] },
+  { rintro ⟨y, hy_fixed⟩,
+    use e.symm y,
+    apply e.injective,
+    simp [hy_fixed] }
+end
+
+end derangement_lemmas
+
+
+section ignore_swap
+
+variables {α : Type*} [decidable_eq α]
+
+/-- A predicate describing if permutation `f` swaps `a` and `b`.
+    Note that `a` and `b` need not be distinct. -/
+def swaps {α : Type*} (f : perm α) (a b : α) : Prop := f a = b ∧ f b = a
+
+/-- The set of permutations that swap `a` and `b` is equivalent to permutations on all other
+    elements. -/
+def perm_ignore_swap (a b : α) : { f : perm α | swaps f a b } ≃ perm ({a, b}ᶜ : set α) :=
+begin
+  let a_and_b : set α := {a, b},
+
+  -- TODO why can't it infer this is decidable?
+  haveI : decidable_pred a_and_b := by
+  { dsimp [decidable_pred, a_and_b],
+    intro x,
+    apply set.decidable_mem },
+
+  let swap : perm α := equiv.swap a b,
+  let swap' : perm a_and_b := perm.subtype_perm swap (by {
+    intro x,
+
+    simp,
+    conv_rhs { rw apply_eq_iff_eq_symm_apply, rw apply_eq_iff_eq_symm_apply },
+    simp [swap, or_comm] }),
+
+  transitivity {f : perm α // ∀ x : a_and_b, f x = swap' x},
+  { apply subtype_equiv_right,
+    intro σ,
+    simp [swaps] },
+  { exact equiv.set.compl swap' },
+end
+
+-- TODO should this be a simp lemma?
+lemma perm_ignore_two_cycle_eq (a b : α) (f : perm α) (hf : swaps f a b) (c : ({a, b}ᶜ : set α)):
+  ((perm_ignore_swap a b) ⟨f, hf⟩ c : α) = f c := by simp [perm_ignore_swap, equiv.set.compl]
+
+-- TODO this name definitely isn't up to snuff... but what would it be called?
+lemma perm_ignore_two_cycle_derangement_iff (a b : α) (hab : a ≠ b) (f : perm α) (hf : swaps f a b) :
+  is_derangement f ↔ is_derangement ((perm_ignore_swap a b) ⟨f, hf⟩) :=
+begin
+  split,
+
+  { rintros f_derangement ⟨x, hx⟩,
+    simp [perm_ignore_swap, equiv.set.compl],
+    exact f_derangement x },
+
+  { -- easier to show that fixed point in f implies a fixed point
+    -- in (f without a↔b)
+    contrapose!,
+    simp [is_derangement],
+
+    --intro our fixed point and show that it's neither a nor b
+    intros x x_fixed,
+    have x_ne_a_b : ¬(x = a ∨ x = b),
+    { intro contra,
+      unfold swaps at hf,
+      cases contra; { rw contra at x_fixed, cc } },
+
+    use [x, x_ne_a_b],
+    rw [subtype.ext_iff, perm_ignore_swap],
+    exact x_fixed }
+end
+
+end ignore_swap
+
+
+section splicing
+
+-- TODO feels like this should already exist? something something equiv_functor?
+-- I tried mul_action.to_perm (perm α) (perm α), but that didn't simp easily
+/-- Post-multiplying by a given `f : perm α` gives a permutation on `perm α`. -/
+def equiv_compose {α : Type*} (f : perm α) : perm (perm α) :=
```

Comment:
This seems like it should at least work for all groups (as right-multiplication), even if `mul_action.to_perm` doesn't work for some reason. Could you point out a few places where `simp` gets mad at `mul_action.to_perm`, so I could try to help troubleshoot that?

Reply:
Sure! Here's what happens when I use `mul_action.to_perm` in place of defining my own. There's a lot of definitions I have to expand before it arrives at "just compose the permutations": https://pastebin.com/32ZhFC55